### PR TITLE
fix(cloudfront): ensures origin groups are added with their own ID as a target

### LINKED ISSUE #9561: [cloudfront] OriginGroup doesn't register correctly as DefaultCacheBehavior target.
When I create a Distribution with a defaultBehavior origin that points to a OriginGroup, the outcoming templates target ID points to the primaryOrigin of that OriginGroup instead of to the OriginGroup itself.

### Reproduction Steps

```
   const originGroup = new OriginGroup({
      primaryOrigin: new S3Origin(someBucket, {originPath: "/"}),
      fallbackOrigin: new HttpOrigin(`${api.restApiId}.execute-api.${this.region}.amazonaws.com`, {originPath: `/${api.deploymentStage.stageName}/thumbnail`}),
      fallbackStatusCodes: [404]}
    )
    const cdn = new Distribution(this, 'cdn', {
      defaultBehavior: {
        origin: originGroup
      },
    })
```

This (and many other configurations) create a Distribution with the primary origins target ID registered as default target, rather than the origin group.

### What did you expect to happen?

I expected that the originGroup would be targeted by default.

### Environment

  - **CLI Version      : ** 1.57.0
  - **Framework Version: ** 1.57.0
  - **Node.js Version:** v14.7.0
  - **OS               :** macOS 10.15.5
  - **Language (Version):** Typescript 3.7.2

### Other

The issue was introduced as part of the original implementation of OriginGroups: https://github.com/aws/aws-cdk/pull/9360

I have a fix ready at https://github.com/aws/aws-cdk/compare/master...LarsFronius:default-behaviour-origin-groups - let me know if this looks as expected and I'll create the PR.

/cc ing @skinny85 : Since you created the original implementation and have the most context, perhaps you are the best to give this a review. Thank you!

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by LarsFronius**
This is to fix the issue I created yesterday https://github.com/aws/aws-cdk/issues/9561 - mostly to get the CI running and see if the test pass already. The whole issue is described in #9561.

Fixes #9561

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by LarsFronius** [comment]
Hi @njlynch - I extended the existing test for the origin groups to accommodate the behavior.
Maybe I got lucky, but I am failing to see the regression you are describing.

The call to find the existingOrigin in the boundOrigins properly returns the origin group, which exposes the groups origin id via the `originId` field for me in the test case I've added.

Maybe we can work together in finding a suitable failing test?

File: packages/@aws-cdk/aws-cloudfront-origins/test/integ.origin-group.expected.json [hunk 1]
```diff
@@ -72,9 +72,20 @@
...
             "ViewerProtocolPolicy": "allow-all"
           },
+          "CacheBehaviors": [
+            {
+              "PathPattern": "/api",
+              "ForwardedValues": {
+                "QueryString": false
+              },
+              "PathPattern": "/api",
+              "TargetOriginId": "cloudfrontorigingroupDistributionOrigin137659A54",
```
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront-origins/test/integ.origin-group.expected.json:85]: This should be `cloudfrontorigingroupDistributionOriginGroup10B57F1D1`, not `cloudfrontorigingroupDistributionOrigin137659A54`, correct?

Comment (LarsFronius) [reply packages/@aws-cdk/aws-cloudfront-origins/test/integ.origin-group.expected.json:85]: Aaah, now I see it. I had assumed it would return an origin instead of an origin group - did not look after the concrete hash. Right!

**Comment by skinny85** [comment]
Hey @LarsFronius ,

thanks for opening the issue. Your build is failing with:

```
@aws-cdk/aws-cloudfront-origins: Verifying integ.origin-group.js against integ.origin-group.expected.json ... CHANGED.
@aws-cdk/aws-cloudfront-origins: Resources
@aws-cdk/aws-cloudfront-origins: [~] AWS::CloudFront::Distribution DistributionCFDistribution882A7313 
@aws-cdk/aws-cloudfront-origins:  â””â”€ [~] DistributionConfig
@aws-cdk/aws-cloudfront-origins:      â””â”€ [~] .CacheBehaviors:
@aws-cdk/aws-cloudfront-origins:          â””â”€ @@ -4,7 +4,7 @@
@aws-cdk/aws-cloudfront-origins:             [ ]       "QueryString": false
@aws-cdk/aws-cloudfront-origins:             [ ]     },
@aws-cdk/aws-cloudfront-origins:             [ ]     "PathPattern": "/api",
@aws-cdk/aws-cloudfront-origins:             [-]     "TargetOriginId": "cloudfrontorigingroupDistributionOrigin10B57F1D1",
@aws-cdk/aws-cloudfront-origins:             [+]     "TargetOriginId": "cloudfrontorigingroupDistributionOrigin137659A54",
```

I guess the `integ.origin-group.expected.json` needs some further editing.

**Comment by skinny85** [comment]
> Thanks for this PR!
> 
> The change you've made looks good, but I think only fixes the cases for a single behavior. Consider a distribution like this:
> 
> ```ts
> new cloudfront.Distribution(stack, 'Distribution', {
>   defaultBehavior: { origin: originGroup },
>   additionalBehaviors: {
>     '/api': {
>       origin: originGroup,
>     },
>   },
> });
> ```
> 
> In the above case, the default behavior will get the proper `originGroupId` as the target, but for the additional behavior the first `originId` will be used. This is because of the `existingOrigin` check:
> 
> https://github.com/aws/aws-cdk/blob/90a705b714cc687ccf01186771f980d53f1f4a70/packages/%40aws-cdk/aws-cloudfront/lib/distribution.ts#L193-L195
> 
> One approach might be to add the `originGroupId` to the `BoundOrigin` interface, and then have the existingOrigin return it if present.
> 
> ```ts
> if (existingOrigin) {
>   return existingOrigin.originGroupId ?? existingOrigin.originId;
> }
> ```

+1 to Nick's suggestion. This is quite tricky with all of the IDs flying around, but I think this is the simplest way to solve this problem.

**Comment by LarsFronius** [comment]
> Hey @LarsFronius ,
> 
> thanks for opening the issue. Your build is failing with:
> 
> ```
> @aws-cdk/aws-cloudfront-origins: Verifying integ.origin-group.js against integ.origin-group.expected.json ... CHANGED.
> @aws-cdk/aws-cloudfront-origins: Resources
> @aws-cdk/aws-cloudfront-origins: [~] AWS::CloudFront::Distribution DistributionCFDistribution882A7313 
> @aws-cdk/aws-cloudfront-origins:  â””â”€ [~] DistributionConfig
> @aws-cdk/aws-cloudfront-origins:      â””â”€ [~] .CacheBehaviors:
> @aws-cdk/aws-cloudfront-origins:          â””â”€ @@ -4,7 +4,7 @@
> @aws-cdk/aws-cloudfront-origins:             [ ]       "QueryString": false
> @aws-cdk/aws-cloudfront-origins:             [ ]     },
> @aws-cdk/aws-cloudfront-origins:             [ ]     "PathPattern": "/api",
> @aws-cdk/aws-cloudfront-origins:             [-]     "TargetOriginId": "cloudfrontorigingroupDistributionOrigin10B57F1D1",
> @aws-cdk/aws-cloudfront-origins:             [+]     "TargetOriginId": "cloudfrontorigingroupDistributionOrigin137659A54",
> ```
> 
> I guess the `integ.origin-group.expected.json` needs some further editing.

That was expected - @njlynch gave me the right hint in what edge case was still not covered and thus went on to create the failing test to make sure it will be covered before the PR gets merged :) 

Unfortunately my exposure with the CDK internals has been fairly limited outside of the issue. I'll have a go at the issue once more today, but if one of you knows exactly what to do feel free to take it away from me. 👍

**Comment by njlynch** [comment]
> I'll have a go at the issue once more today, but if one of you knows exactly what to do feel free to take it away from me.

Keep going, you're almost there! :)

I think the best bet is to add a new property to the `BoundOrigin` interface:

https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-cloudfront/lib/distribution.ts#L58-L60

Something like this:

```
interface BoundOrigin extends OriginBindOptions, OriginBindConfig {
  readonly origin: IOrigin;
  readonly originGroupId: string;
}
```

And use this in the `addOrigin` method to add the origin group ID to the entry as it's added to `this.boundOrigins`.

https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-cloudfront/lib/distribution.ts#L300

File: packages/@aws-cdk/aws-cloudfront-origins/test/origin-group.test.ts [hunk 2]
```diff
@@ -28,8 +28,12 @@ describe('Origin Groups', () => {
 
     const primaryOriginId = 'DistributionOrigin13547B94F';
     const failoverOriginId = 'DistributionOrigin2C85CC43B';
+    const originGroupId = 'DistributionOriginGroup1A1A31B49';
     expect(stack).toHaveResourceLike('AWS::CloudFront::Distribution', {
       DistributionConfig: {
+        DefaultCacheBehavior: {
+	        TargetOriginId: originGroupId,
```
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront-origins/test/origin-group.test.ts:35]:
```suggestion
          TargetOriginId: originGroupId,
```

File: packages/@aws-cdk/aws-cloudfront/lib/distribution.ts [hunk 3]
```diff
@@ -291,38 +292,43 @@ export class Distribution extends Resource implements IDistribution {
...
-      return existingOrigin.originId;
+      return existingOrigin.originGroupId ?? existingOrigin.originId;
     } else {
       const originIndex = this.boundOrigins.length + 1;
       const scope = new Construct(this, `Origin${originIndex}`);
       const originId = scope.node.uniqueId;
       const originBindConfig = origin.bind(scope, { originId });
-      this.boundOrigins.push({ origin, originId, ...originBindConfig });
-      if (originBindConfig.failoverConfig) {
+      if (! originBindConfig.failoverConfig) {
```
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront/lib/distribution.ts:301]:
```suggestion
      if (!originBindConfig.failoverConfig) {
```

**Comment by LarsFronius** [comment]
> Looks great, thanks!

Thanks for working with me through this!
It's been a bit odd navigating the quite unfamiliar codebase and making sense of the functionality (even if I understand CloudFormation really well 😅 ), so much appreciate all the guidance you gave getting this over the finish line!
